src/Chainweb/Pact5/Backend/ChainwebPactDb.hs

```diff
@@ -594,8 +594,14 @@ doKeys d = do
               Just v -> pure (v, Dict ())
               Nothing -> internalDbError $ "doKeys.DModuleSources: unexpected decoding"
     case ordDict of
-        Dict () ->
-            return $ sort (memKeys ++ parsedKeys)
+        Dict () -> do
+            v <- view blockHandlerVersion
+            cid <- view blockHandlerChainId
+            bh <- view blockHandlerBlockHeight
+            if chainweb230Pact v cid bh
+            -- the read-cache contains duplicate keys that we need to remove.
+            then return $ fmap head $ group $ sort (memKeys ++ parsedKeys)
```

Comment:
use unsafeHead to avoid GHC warnings about partial functions

Reply:
```suggestion
            then return $ fmap unsafeHead $ group $ sort (memKeys ++ parsedKeys)
```